pennylane_lightning/lightning_qubit/lightning_qubit.py

```diff
@@ -115,6 +116,55 @@ def simulate_and_jacobian(circuit: QuantumTape, state: LightningStateVector, bat
     return res, jac
 
 
+def vjp(
+    circuit: QuantumTape, cotangents: Tuple[Number], state: LightningStateVector, batch_obs=False
+):
+    """Compute the Vector-Jacobian Product (VJP) for a single quantum script.
+    Args:
+        circuit (QuantumTape): The single circuit to simulate
+        cotangents (Tuple[Number, Tuple[Number]]): Gradient-output vector. Must
+            have shape matching the output shape of the corresponding circuit. If
+            the circuit has a single output, ``cotangents`` may be a single number,
+            not an iterable of numbers.
+        state (LightningStateVector): handle to Lightning state vector
+        batch_obs (bool): Determine whether we process observables in parallel when
+            computing the VJP. This value is only relevant when the lightning
+            qubit is built with OpenMP.
+    Returns:
+        TensorLike: The VJP of the quantum script
+    """
+    circuit = circuit.map_to_standard_wires()
+    state.reset_state()
+    final_state = state.get_final_state(circuit)
```

Comment:
Given this could potentially be expensive, we could potentially keep track of the `tape.hash` that corresponds to the current state, and only reseting and recalculating the state when the hash is different.

Reply:
@albi3ro, you mean that calculating the final state is expensive, right?
I think it would be a smart move to hash in order to avoid re-calculating. However, wouldn't it make more sense to hash `tape.operations` instead? This is because we essentially use `tape.operations` to build the state vector, and even if the observables change, the state vector would still be the same. The same logic applies to both the simulate and jacobian parts, right?

Reply:
I'm not sure how significant the benefit would be for caching using `tape.operations` hash vs `tape.hash`. However, if we consider doing this, same change will also benefit `default.qubit`.

Reply:
I think we can let this for now, and come back in the future if needed.